For migration to 3, accommodate case companion as function

som-snytt opened the pull request:
Scala 2 sometimes makes the companion object of a case class extend `Function`; Scala 3 does not, but includes other adaptations which are backported under `-Xsource:3-cross` and allow the following idioms with reduced ceremony:

```
case class B(i: Int)
case class C(i: Int, j: Int)
def mapped(xs: List[Int]): List[B] = xs.map(B)
def cross(xs: List[(Int, Int)]): List[C] = xs.map(C)
def f(xs: List[(Int, Int)]): List[C] = xs.map(C.tupled)
def g(xs: List[Int]): List[C] = xs.map(C.curried).map(_(42))
```

Under `-Xsource:3`, using the case class companion as a function will warn.

Backports Dotty migration support of case class companions by inserting `apply`. Dotty does not automatically add `Function` as a parent of the companion; for Scala 2 under `-Xsource:3-cross`, follow Dotty but also adapt where the companion is not already a `Function`.

Since Scala 2 lacks parameter untupling, also support explicit `C.tupled` and for completeness `C.curried`.

Fixes scala/bug#3664

som-snytt committed:
Warn synthetic case object used as function

som-snytt committed:
Adapt apply/tupled/curried insertion for case companion

joroKr21 commented:
Oh wait, why add a feature that is deprecated right away? 🤔

som-snytt commented:
Not sure about knobs for enablement yet, but Scala 3 warns about apply insertion. It's strictly a migration mitigation.

som-snytt commented:
blocked by https://github.com/scala/scala/pull/10573

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -1122,12 +1126,20 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
+      def adaptApplyInsertion(): Tree = {
+        val msg = s"The method `apply` is inserted. The auto insertion will be deprecated, please write `${tree.symbol.name}.apply` explicitly."
+        context.deprecationWarning(tree.pos, tree.symbol, msg, "2.13.13")
+        typed(Select(tree, nme.apply), mode, pt)
```
```suggestion
        typed(atPos(tree.pos)(Select(tree, nme.apply)), mode, pt)
```

som-snytt replied:
I forgot how to make trees. Also, I never got to use `q"uotes"` much. On my other PR, I was challenged to allocate an array.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -5360,6 +5372,22 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
+      def fixUpCaseTupled(qual: Tree, name: Name, mode: Mode): Tree =
+        if (qual.symbol != null && isScala3cross && qual.symbol.isModule && qual.symbol.companion.isCase &&
+            context.undetparams.isEmpty && fixableFunctionMembers.contains(name)) {
+          val t = Select(qual, nme.apply)
```
could also add `atPos` in the trees created here

lrytz reviewed (commented):
LGTM 👍  thanks!

som-snytt commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -5646,9 +5679,12 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
-          }
+            }
+            if (!isPastTyper && currentRun.isScala3 && !isScala3cross && isFunctionType(pt) && sym.isModule && sym.isSynthetic && sym.companion.isCase)
+              context.deprecationWarning(tree.pos, sym, s"Synthetic case companion used as a Function, use explicit object with Function parent", "2.13.13")
```
The message could suggest `-Xsource:3cross`? The end goal also in Scala 3 is explicit, either explicit parent or explicit `C.apply`. 🤔

joroKr21 commented:
> Not sure about knobs for enablement yet, but Scala 3 warns about apply insertion. It's strictly a migration mitigation.

`C.apply` works both in Scala 2 and Scala 3 right?

som-snytt commented:
@joroKr21 iirc the tupled case is not handled because scala 2 doesn't do parameter untupling adaptation. I'll try to nail it down while polishing.

The other missing parts are compose & andThen, so the adaptation here is incomplete in that sense. So really, we're just trying to make it easier for those slick users. I used slick once, do I have to try it again to test out the use case? 😿

som-snytt commented:
@lrytz about to contribute a bit of doc for the incompatibility matrix as you pointed out.

I haven't tried it, but maybe it makes sense to support `case class D(i: Int, j: Int); List(42->27).map(D)` directly, by injection of `tupled`, where the syntax is the same as Scala 3 but the mechanism is slightly different.

lrytz commented:
> maybe it makes sense to support `case class D(i: Int, j: Int); List(42->27).map(D)` directly

Since this doesn't currently work on Scala 2, and only works in Scala 3 with a warning and is therefore discouraged, I don't think this is needed.

som-snytt commented:
My case for `tupled` injection is cross-compilation. For the tupled case, it requires different sources because `D.tupled` is not supported by Scala 3. I haven't tried slick use case yet, but it's the tupled case they care about. I assume slick supports Scala 3.

lrytz commented:
OK, good point

som-snytt commented:
Not sure I'll get around to tupled.

SethTisue commented:
@som-snytt I've provisionally moved this to the 2.13.14 milestone, but maybe it could still make 2.13.13.

It looks to me like there are some unaddressed comments from Lukas about the details of the code.

The rest of the discussion is a bit difficult to follow and I'm not sure where things stand. When you believe this is as ready for merge as you know how to make it, please say so.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -5360,6 +5370,24 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
+            typed(atPos(tree.pos)(Select(etaExpand(t1, context.owner), name)), mode, pt)
+          }
+          if (!t2.isErroneous) {
+            val msg = s"The method `apply` is inserted. The auto insertion will be deprecated, please write `${t2}` explicitly."
```
```suggestion
            val msg = s"The method `apply` is inserted. The auto insertion will be deprecated, please write `($qual.apply _).$name` explicitly."
```

lrytz replied:
`C.apply.tupled` would be nice but that's a different backport.

som-snytt replied:
I believe I have a commit to push.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -1122,12 +1124,20 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           if (settings.isDebug && settings.explaintypes.value) explainTypes(tree.tpe, pt)
           if (err ne null) context.issue(err)
           if (tree.tpe.isErroneous || pt.isErroneous) setError(tree)
+          else if (currentRun.isScala3Cross && isFunctionType(pt) && tree.symbol.isModule && tree.symbol.companion.isCase)
```
`tree.symbol` is null in `neg/t12919-3cross.scala`.

lrytz reviewed (commented):
LGTM once the details are worked out

som-snytt commented:
Will follow up again later today. Helpful pushes turned off for now.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -5360,6 +5370,24 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
+            val t1 = typedSelect(t, qual, nme.apply)
+            typed(atPos(tree.pos)(Select(etaExpand(t1, context.owner), name)), mode, pt)
+          }
+          if (!t2.isErroneous) {
```
I was going to suggest using `silent` above, but actually the error message is helpful.

```
➜ sandbox qs -Xsource:3-cross
scala> case class D(x: Int, y: Int, z: Int)
scala> Some(1 -> 1).map(D.tupled)
                          ^
       error: type mismatch;
        found   : ((Int, Int, Int)) => D
        required: ((Int, Int)) => ?
```

som-snytt replied:
added a test

lrytz reviewed (approved):
This LGTM, let me know if you want to add anything or we can merge.

som-snytt committed:
Adapt tupled apply insertion for case companion

som-snytt commented:
@lrytz added a commit for "insert apply with tupling adaptation". It's a bit tricky (subtle), so it is unsquashed.

I re-enabled the lrytz maneuver for your convenience. (Feel free to push here again.)

Perhaps there is better API usage for the simple questions in `adaptApplyInsertion`, such as "does it have an apply method and does its arity match the expected type or the tuple param", etc.

Slightly augmented tests. My takeaway vis-a-vis TDD is that for a code change of a certain complexity, discipline is required to add tests for intersecting cases. I don't evince that discipline here. A nice IDE tool would be a test generator that probes feature interactions, scalacheck + fuzzing + chatgpt. It's nice to have the community build, but we could use the corpus to generate targeted tests (because we know how people use case classes and tupled, for instance). Oh, I just thought of another test.

som-snytt committed:
Apply insertion for nilary case companion

lrytz commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -1122,13 +1124,39 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
+                case args if args.lengthCompare(arity) == 0 =>
+                  val msg = s"The method `apply` is inserted. The auto insertion will be deprecated, please write `${tree.symbol.name}.apply` explicitly."
+                  context.deprecationWarning(tree.pos, tree.symbol, msg, "2.13.13")
+                  typed(atPos(tree.pos)(Select(tree, nme.apply)), mode, pt)
```
```suggestion
                  applicator
```
?

som-snytt replied:
Gee, I thought I'd tried all the combinations of supplying pt. Using applicator works. So the silent doesn't get pt, and I didn't drill into why that gives me the function; the caveat is that `case class E()` is not expanded; supplying pt fixes that but then the tupled case errors right away. The fix then is if arity < 0, retype with pt for E. I'll add the test for E and also for overloaded apply, which should error nicely. I checked that dotty does apply insert for E. (I understand scala 2 resists eta expanding nilary method.)

som-snytt commented:
Rebased and added support for `case class E()` (which is adapted by dotty).

The first typecheck of `C.apply` uses pt (so E works). On error, use wildcard (so tupled works).

Note: using wildcard first emits spurious warning about "empty application" of `E()`. Even though the result is discarded by us, the successful result makes silent emit the warnings. (Possibly you could `typer.context.reporter.clearAll()` before returning from the silent op, but it's clunky.)

som-snytt commented:
This is like the old days when CI runs would just fail periodically. Is this part of the "scripted" tests? It's after partest.

```
[info] [success] Total time: 0 s, completed Feb 2, 2024 6:47:09 PM
[info] [info] welcome to sbt 1.9.8 (Oracle Corporation Java 1.8.0_332)
[info] [info] loading project definition from /tmp/sbt_31b28e22/project
[info] [info] loading settings for project macro-provider from common.sbt ...
[info] [info] loading settings for project macro-client from common.sbt ...
[info] [info] loading settings for project sbt_31b28e22 from build.sbt,common.sbt ...
[info] [info] set current project to sbt_31b28e22 (in build file:/tmp/sbt_31b28e22/)
[info] [info] compiling 1 Scala source to /tmp/sbt_31b28e22/macro-provider/target/classes ...
[info] [error] ## Exception when compiling 1 sources to /tmp/sbt_31b28e22/macro-provider/target/classes
[info] [error] java.lang.NoClassDefFoundError: scala/reflect/macros/internal/macroImpl
[info] [error] scala.reflect.internal.Definitions$DefinitionsClass.MacroImplAnnotation$lzycompute(Definitions.scala:587)
[info] [error] scala.reflect.internal.Definitions$DefinitionsClass.MacroImplAnnotation(Definitions.scala:587)
[info] [error] scala.tools.nsc.typechecker.Macros.$anonfun$loadMacroImplBinding$1(Macros.scala:291)
[info] [error] scala.collection.mutable.AnyRefMap.getOrElseUpdate(AnyRefMap.scala:169)
[info] [error] scala.tools.nsc.typechecker.Macros.loadMacroImplBinding(Macros.scala:293)
[info] [error] scala.tools.nsc.typechecker.Macros.loadMacroImplBinding$(Macros.scala:289)
[info] [error] scala.tools.nsc.Global$$anon$6.loadMacroImplBinding(Global.scala:512)
[info] [error] scala.tools.nsc.typechecker.Macros.standardIsBlackbox(Macros.scala:310)
```

som-snytt commented:
/rebuild

SethTisue commented:
wasn't sure if `/rebuild` does older commits (the ones that the "combined" check is complaining about), but I see at https://scala-ci.typesafe.com/job/scala-2.13.x-validate-main/ that it does

som-snytt commented:
I must have broken something when I rebased? It was green before. 😿 

Oh I guess just warning about something could induce a promoted error. But it's a totally different stack trace.
```
[info] [info] set current project to sbt_57597d91 (in build file:/tmp/sbt_57597d91/)
[info] [success] Total time: 0 s, completed Feb 3, 2024 5:08:29 PM
[info] [info] welcome to sbt 1.9.8 (Oracle Corporation Java 1.8.0_332)
[info] [info] loading project definition from /tmp/sbt_57597d91/project
[info] [info] loading settings for project sbt_57597d91 from common.sbt ...
[info] [info] set current project to sbt_57597d91 (in build file:/tmp/sbt_57597d91/)
[info] [info] compiling 2 Scala sources and 1 Java source to /tmp/sbt_57597d91/target/classes ...
[info] [error] ## Exception when compiling 3 sources to /tmp/sbt_57597d91/target/classes
[info] [error] java.lang.NoClassDefFoundError: scala/runtime/RichInt
[info] [error] scala.reflect.internal.Phase.compare(Phase.scala:73)
[info] [error] scala.reflect.internal.Phase.compare(Phase.scala:17)
[info] [error] scala.math.Ordered.$greater(Ordered.scala:82)
[info] [error] scala.math.Ordered.$greater$(Ordered.scala:82)
[info] [error] scala.reflect.internal.Phase.$greater(Phase.scala:17)
[info] [error] scala.tools.nsc.Global$GlobalPhase.shouldSkipThisPhaseForJava(Global.scala:435)
[info] [error] scala.tools.nsc.Global$GlobalPhase.cancelled(Global.scala:440)
[info] [error] scala.tools.nsc.Global$GlobalPhase.applyPhase(Global.scala:476)
[info] [error] scala.tools.nsc.Global$GlobalPhase.run(Global.scala:428)
[info] [error] scala.tools.nsc.Global$Run.compileUnitsInternal(Global.scala:1549)
[info] [error] scala.tools.nsc.Global$Run.compileUnits(Global.scala:1533)
[info] [error] scala.tools.nsc.Global$Run.compileSources(Global.scala:1525)
[info] [error] scala.tools.nsc.Global$Run.compileFiles(Global.scala:1638)
[info] [error] scala.tools.xsbt.CachedCompiler0.run(CompilerBridge.scala:176)
[info] [error] scala.tools.xsbt.CachedCompiler0.run(CompilerBridge.scala:139)
[info] [error] scala.tools.xsbt.CompilerBridge.run(CompilerBridge.scala:43)
```

som-snytt commented:
/rebuild

lrytz reviewed (approved):
Very nice! Sorry for the CI overhead...

SethTisue commented:
@som-snytt can you do a round of making the PR description user-facing here? I think it would be valuable to include an example, and to explicitly say what the behavior is on `-Xsource:3` vs `-Xsource:3-cross` vs neither.

som-snytt commented:
> I'll try to nail it down while polishing.

The pun was 💅 .

lrytz commented:
9 months and noone noticed...
